### PR TITLE
feat: Add template settings page

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Related to #2221 

Demo:

https://user-images.githubusercontent.com/3165839/185460193-25468045-b5ed-4722-af8a-eb12136ac1de.mov

### CODE REVIEW

File: site/src/xServices/templateSettings/templateSettingsXService.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      services: {
+        getTemplateSettings: async ({ organizationId, templateName }) => {
+          return getTemplateByName(organizationId, templateName)
+        },
+        saveTemplateSettings: async (
+          { templateSettings },
+          { templateSettings: newTemplateSettings },
+        ) => {
+          if (!templateSettings) {
+            throw new Error("templateSettins is not loaded yet.")
```
Comment (presleyp) [inline site/src/xServices/templateSettings/templateSettingsXService.ts:86]:
```suggestion
            throw new Error("templateSettings is not loaded yet.")
```

File: site/src/pages/TemplateSettingsPage/TemplateSettingsPage.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  const organizationId = useOrganizationId()
+  const [state, send] = useMachine(templateSettingsMachine, {
+    context: { templateName, organizationId },
+    actions: {
+      onSave: (_, { data }) => {
+        // Use the data.name because the template name can be changed
+        navigate(`/templates/${data.name}`)
+      },
+    },
+  })
```
Comment (presleyp) [inline site/src/pages/TemplateSettingsPage/TemplateSettingsPage.tsx:26]: nice!

**Review by kylecarbs — APPROVED** [review]
I think the full-page modal design will bite us in the relatively short-term future. When we add more options, the page will only get longer and eventually have to be recreated (probably even if we added 4-5 more).

image

I like the hierarchy of design here from Vercel. It displays the resource being edited (in this case our coder.com deployment) and has a scalable design for adding more options. The design in this PR is functional for now but doesn't have the professional polish I think would be nice. That is nice to have though, so I certainly won't block this change!

(I don't have time to do a code review right now, I just watched the demo, so someone else should review!)

**Review by presleyp — APPROVED** [review]
The code looks good and chromatic mostly looks good - I didn't see any error stuff on the save error message story so I denied it in Chromatic just so you can take a look.

**Comment by BrunoQuaresma** [comment]
> The code looks good and chromatic mostly looks good - I didn't see any error stuff on the save error message story so I denied it in Chromatic just so you can take a look.

Yes, it is because you have to click outside the form... Idk why this happens, but it looks like MUI doesn't display the error if it is focused - the input is auto-focus. Not sure if I should add some auto-focus logic to enable it only when the form doesn't have errors... sounds too much. What do you think @presleyp ?

**Comment by BrunoQuaresma** [comment]
@kylecarbs I agree with you. I just would expect we have more options because a page like the Vercel one, only looks good if you have options enough to fill the space.

**Pull request merged**